packages/hop-node/src/1inch/1inch.ts

```diff
@@ -0,0 +1,347 @@
+      }
+
+      amount = (await token.getBalance()).toString()
+      formattedAmount = Number(formatUnits(amount.toString(), getTokenDecimals(fromToken)))
+    }
+  } else {
+    amount = parseUnits(formattedAmount.toString(), getTokenDecimals(fromToken)).toString()
+  }
+
+  logger.debug('chain:', chain)
+  logger.debug('fromToken:', fromToken)
+  logger.debug('toToken:', toToken)
+  logger.debug('amount:', formattedAmount)
+  logger.debug('slippage:', slippage)
+  logger.debug('dryMode:', !!dryMode)
+
+  const canonicalFromTokenSymbol = getCanonicalTokenSymbol(fromToken)
+  const canonicalToTokenSymbol = getCanonicalTokenSymbol(toToken)
+  const fromTokenConfig = (mainnetAddresses as any).bridges?.[canonicalFromTokenSymbol]?.[chain]
+  const toTokenConfig = (mainnetAddresses as any).bridges?.[canonicalToTokenSymbol]?.[chain]
+  let fromTokenAddress = fromTokenConfig?.l1CanonicalToken || fromTokenConfig?.l2CanonicalToken
+  let toTokenAddress = toTokenConfig?.l1CanonicalToken || toTokenConfig?.l2CanonicalToken
+
+  // 1inch uses 0xeee…eee address for native tokens
+  if (isFromNative) {
+    fromTokenAddress = '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee'
+  }
+  if (isToNative) {
+    toTokenAddress = '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee'
+  }
+
+  if (!fromTokenAddress) {
+    throw new Error(`from token "${fromToken}" is not supported`)
+  }
+
+  if (!toTokenAddress) {
+    throw new Error(`to token "${toToken}" is not supported`)
+  }
+
+  logger.debug('checking allowance')
+  const tokenAddress = fromTokenAddress
+  const allowance = await oneInch.getAllowance({ tokenAddress, walletAddress })
+  logger.debug('allowance:', formatUnits(allowance, getTokenDecimals(fromToken)))
+
+  if (BigNumber.from(allowance).lt(amount)) {
+    const txData = await oneInch.getApproveTx({ tokenAddress, amount })
+    logger.debug('approval data:', txData)
+
+    if (dryMode) {
+      logger.debug('dryMode enabled, skipping approve tx')
+    } else {
+      logger.debug('sending approve tx')
+      const tx = await wallet.sendTransaction(txData)
+      logger.debug('approval tx:', tx.hash)
+      await tx.wait()
+    }
+  }
+
+  logger.debug('checking quote')
+  const toTokenAmount = await oneInch.getQuote({ fromTokenAddress, toTokenAddress, amount })
+  const toTokenAmountFormatted = formatUnits(toTokenAmount, getTokenDecimals(toToken))
+  logger.debug(`toTokenAmount: ${toTokenAmountFormatted}`)
+
+  logger.debug('getting swap data')
+  const fromAddress = walletAddress
+  const txData = await oneInch.getSwapTx({ fromTokenAddress, toTokenAddress, fromAddress, amount, slippage, destReceiver: recipient })
+  logger.debug('swap data:', txData)
+
+  if (dryMode) {
+    logger.debug('dryMode enabled, skipping swap tx')
+  } else {
+    logger.debug('sending swap tx')
+    const tx = await wallet.sendTransaction(txData)
+    logger.debug('swap tx:', tx.hash)
+    return tx
+  }
+}
+
+const nativeChainTokens: Record<string, string> = {
```

Comment:
Nitpick: maybe move this to `constants`. I think I've needed this in a few other places.

Reply:
good idea, updated with https://github.com/hop-protocol/hop/commit/a38b614cc480ac4cfdd492433b6bb30041374d73